plugin/kotel/tracing.go

```diff
@@ -0,0 +1,162 @@
+package kotel
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+
+	"github.com/twmb/franz-go/pkg/kgo"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/codes"
+	"go.opentelemetry.io/otel/propagation"
+	semconv "go.opentelemetry.io/otel/semconv/v1.12.0"
+	"go.opentelemetry.io/otel/trace"
+)
+
+var ( // interface checks to ensure we implement the hooks properly
+	_ kgo.HookProduceRecordBuffered   = new(Tracer)
+	_ kgo.HookProduceRecordUnbuffered = new(Tracer)
+	_ kgo.HookFetchRecordBuffered     = new(Tracer)
+)
+
+type Tracer struct {
+	tracerProvider trace.TracerProvider
+	propagators    propagation.TextMapPropagator
+	tracer         trace.Tracer
+}
+
+// TracingOption interface used for setting optional config properties.
+type TracingOption interface {
+	apply(*Tracer)
+}
+
+type tracingOptionFunc func(*Tracer)
+
+func (o tracingOptionFunc) apply(t *Tracer) {
+	o(t)
+}
+
+// NewTracer returns a Tracer, used as option for kotel to instrument franz-go with tracing
+func NewTracer(opts ...TracingOption) *Tracer {
+	t := &Tracer{}
+	for _, opt := range opts {
+		opt.apply(t)
+	}
+	if t.tracerProvider == nil {
+		t.tracerProvider = otel.GetTracerProvider()
+	}
+	if t.propagators == nil {
+		t.propagators = otel.GetTextMapPropagator()
+	}
+	t.tracer = t.tracerProvider.Tracer(
+		instrumentationName,
+		trace.WithInstrumentationVersion(SemVersion()),
+		trace.WithSchemaURL(semconv.SchemaURL),
+	)
+	return t
+}
+
+// TracerProvider takes a trace.TracerProvider and applies it to the Tracer
+// If none is specified, the global provider is used.
+func TracerProvider(provider trace.TracerProvider) TracingOption {
+	return tracingOptionFunc(func(t *Tracer) {
+		if t != nil {
+			t.tracerProvider = provider
+		}
+	})
+}
+
+// TracerPropagator takes a propagation.TextMapPropagator and applies it to the Tracer
+// If none is specified, the global Propagator is used.
+func TracerPropagator(propagator propagation.TextMapPropagator) TracingOption {
+	return tracingOptionFunc(func(t *Tracer) {
+		if t != nil {
+			t.propagators = propagator
+		}
+	})
+}
+
+// Hooks ---------------------------------------------------------------------
+
+// OnProduceRecordBuffered Begins the producer span
+func (t *Tracer) OnProduceRecordBuffered(r *kgo.Record) {
+	attrs := []attribute.KeyValue{
+		semconv.MessagingSystemKey.String("kafka"),
+		semconv.MessagingDestinationKindTopic,
+		semconv.MessagingDestinationKey.String(r.Topic),
+		semconv.MessagingKafkaClientIDKey.String(strconv.FormatInt(r.ProducerID, 10)),
```

Comment:
Is this meant to be the actual client ID? Looking at the otel docs, I think so -- if so, the client ID can / should be an option to the tracer

(I admit it is ugly that you'll need to specify the client ID in two locations)

Reply:
If we are to proceed with this approach, we should also consider adding the consumer group name and any other necessary information to the options to fully comply with the messaging specification outlined in [1]. This will ensure consistency and compliance with the standard. 

[1] https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md#apache-kafka